### PR TITLE
#CU-83x1f5 dev: Create Autocomplete Component

### PR DESCRIPTION
**Pull request by thienduc12311**

### CODE REVIEW

**Comment by binh1298** [comment]
Task linked: [CU-83x1f5 AutocompleteBase Component](https://app.clickup.com/t/83x1f5)

**Comment by thienduc12311** [comment]
Screen Shot 2021-05-14 at 17 15 30

File: src/components/Autocomplete/AutocompleteBase/index.tsx [hunk 1]
```diff
@@ -0,0 +1,18 @@
...
+    AutocompleteProps as MaterialAutocompleteProps,
+} from "@material-ui/lab";
+
+interface Options {
+    title?: string;
+    description?: string;
+}
+
+export interface AutocompleteBaseProps
+    extends MaterialAutocompleteProps<Options, true, true, true> {}
```
Comment (binh1298) [inline src/components/Autocomplete/AutocompleteBase/index.tsx:12]: what is true true true here?

File: src/components/Autocomplete/AutocompleteBase/index.tsx [hunk 2]
```diff
@@ -0,0 +1,13 @@
+import {
+    Autocomplete as MaterialAutocomplete,
+    AutocompleteProps as MaterialAutocompleteProps,
+} from "@material-ui/lab";
+
+export interface AutocompleteBaseProps
+    extends MaterialAutocompleteProps<Object, boolean, boolean, boolean> {}
```
Comment (binh1298) [inline src/components/Autocomplete/AutocompleteBase/index.tsx:7]:
```suggestion
export interface AutocompleteBaseProps<T>
    extends MaterialAutocompleteProps<T, boolean, boolean, boolean> {}
```
Please look into how typescript generic work. Your current implementation will make the component loses the ability to use generic type
https://www.typescriptlang.org/docs/handbook/2/generics.html#hello-world-of-generics

Comment (binh1298) [reply src/components/Autocomplete/AutocompleteBase/index.tsx:7]: My suggestion is just an example. I think it's also possible to generic the boolean variables

File: src/components/Autocomplete/AutocompleteBase/AutocompleteBase.stories.tsx [hunk 3]
```diff
@@ -0,0 +1,35 @@
...
+    { title: "Binh Pham" },
+    { title: "Thanh Binh" },
+];
+
+export default {
+    title: "Components/AutocompleteBase",
+    component: AutocompleteBase,
+};
+
+const Template: Story<AutocompleteBaseProps<Object, boolean, boolean, boolean>> = (args) => {
```
Comment (binh1298) [inline src/components/Autocomplete/AutocompleteBase/AutocompleteBase.stories.tsx:18]: This is a generic example:
```suggestion
const Template: Story<AutocompleteBaseProps<{title: string}, boolean, boolean, boolean>> = (args) => {
```

File: src/components/Autocomplete/AutocompleteBase/AutocompleteBase.stories.tsx [hunk 4]
```diff
@@ -0,0 +1,35 @@
...
+    return <AutocompleteBase {...args} />;
+};
+export const Basic = Template.bind({});
+Basic.args = {
+    options: options,
+    getOptionLabel: (option: any) => option.title,
```
Comment (binh1298) [inline src/components/Autocomplete/AutocompleteBase/AutocompleteBase.stories.tsx:24]: This is why you need generic. If you do it right, the option will automatically get its type 
```suggestion
    getOptionLabel: (option) => option.title,
```

File: src/components/Autocomplete/AutocompleteBase/AutocompleteBase.stories.tsx (hunk 3 above)
Comment (binh1298) [reply src/components/Autocomplete/AutocompleteBase/AutocompleteBase.stories.tsx:18]: You are almost there, good luck

File: src/components/Autocomplete/AutocompleteBase/AutocompleteBase.stories.tsx [hunk 5]
```diff
@@ -0,0 +1,37 @@
...
+    options: options,
+    getOptionLabel: (option) => option.title,
+    renderInput: (params) => <TextField {...params} label="Name" variant="outlined" />,
+    style: { width: 300 },
+};
+export const MultiValue = Template.bind({});
+MultiValue.args = {
+    multiple: true,
+    options: options,
+    getOptionLabel: (option: any) => option.title,
```
Comment (binh1298) [inline src/components/Autocomplete/AutocompleteBase/AutocompleteBase.stories.tsx:34]:
```suggestion
    getOptionLabel: (option) => option.title,
```

File: src/components/Autocomplete/AutocompleteBase/AutocompleteBase.stories.tsx [hunk 6]
```diff
@@ -0,0 +1,37 @@
...
+    { title: "Binh Pham" },
+    { title: "Thanh Binh" },
+];
+
+export default {
+    title: "Components/AutocompleteBase",
+    component: AutocompleteBase,
+};
+
+const Template: Story<AutocompleteBaseProps<{ title: string }, boolean, boolean, boolean>> = (
```
Comment (binh1298) [inline src/components/Autocomplete/AutocompleteBase/AutocompleteBase.stories.tsx:18]:
```suggestion
interface UserInfo {
  title: string
}
const options: UserInfo[] = [
    { title: "Thien Duc" },
    { title: "Kien Tran" },
    { title: "Binh Pham" },
    { title: "Thanh Binh" },
];

export default {
    title: "Components/AutocompleteBase",
    component: AutocompleteBase,
};

const Template: Story<AutocompleteBaseProps<UserInfo, boolean, boolean, boolean>> = (
```

File: src/components/Autocomplete/AutocompleteBase/index.tsx [hunk 7]
```diff
@@ -0,0 +1,24 @@
...
+    Multiple extends boolean | undefined,
+    DisableClearable extends boolean | undefined,
+    FreeSolo extends boolean | undefined
+> extends MaterialAutocompleteProps<T, Multiple, DisableClearable, FreeSolo> {}
+
+const AutoCompleteBase = <
+    T extends {},
+    Multiple extends boolean | undefined,
+    DisableClearable extends boolean | undefined,
+    FreeSolo extends boolean | undefined
```
Comment (binh1298) [inline src/components/Autocomplete/AutocompleteBase/index.tsx:17]:
```suggestion
    T extends {},
    Multiple extends boolean | undefined = false,
    DisableClearable extends boolean | undefined = false,
    FreeSolo extends boolean | undefined = false
```

File: src/components/Autocomplete/AutocompleteBase/index.tsx [hunk 8]
```diff
@@ -0,0 +1,24 @@
+import {
+    Autocomplete as MaterialAutocomplete,
+    AutocompleteProps as MaterialAutocompleteProps,
+} from "@material-ui/lab";
+
+export interface AutocompleteBaseProps<
+    T,
+    Multiple extends boolean | undefined,
+    DisableClearable extends boolean | undefined,
+    FreeSolo extends boolean | undefined
```
Comment (binh1298) [inline src/components/Autocomplete/AutocompleteBase/index.tsx:10]: T extends {},
    Multiple extends boolean | undefined = false,
    DisableClearable extends boolean | undefined = false,
    FreeSolo extends boolean | undefined = false

File: src/components/Autocomplete/AutocompleteBase/index.tsx [hunk 9]
```diff
@@ -0,0 +1,24 @@
...
+> extends MaterialAutocompleteProps<T, Multiple, DisableClearable, FreeSolo> {}
+
+const AutoCompleteBase = <
+    T extends {},
+    Multiple extends boolean | undefined,
+    DisableClearable extends boolean | undefined,
+    FreeSolo extends boolean | undefined
+>(
```
Comment (binh1298) [inline src/components/Autocomplete/AutocompleteBase/index.tsx:18]:
```suggestion
const AutoCompleteBase = <AutocompleteBaseProps>(
```

Comment (binh1298) [reply src/components/Autocomplete/AutocompleteBase/index.tsx:18]: I'm not sure about this one but please try to not have duplicate code in one file

**Review by binh1298 — APPROVED** [review]

**Pull request merged**